### PR TITLE
gpu: Move Descriptor Binding Layout LUT to CPU

### PR DESCRIPTION
**Pull request by spencer-lunarg**
follow-up of https://github.com/KhronosGroup/Vulkan-ValidationLayers/pull/8714

This moves the LUT from being in the GPU to being on the CPU. The table is basic and can be calculated simply.

### CODE REVIEW

File: layers/gpu/debug_printf/debug_printf.cpp [hunk 1]
```diff
@@ -216,7 +216,7 @@ void AnalyzeAndGenerateMessage(Validator &gpuav, VkCommandBuffer command_buffer,
         }
 
         std::vector<Instruction> instructions;
-        spirv::GenerateInstructions(instrumented_shader->instrumented_spirv, instructions);
+        ::spirv::GenerateInstructions(instrumented_shader->instrumented_spirv, instructions);
```
Comment (spencer-lunarg) [inline layers/gpu/debug_printf/debug_printf.cpp:219]: found some compilers are confused, not sure why, but it is only in 2 spots so can look into this later

File: layers/gpu/descriptor_validation/gpuav_descriptor_validation.cpp [hunk 2]
```diff
@@ -92,12 +92,12 @@ void UpdateBoundDescriptors(Validator &gpuav, CommandBuffer &cb_state, VkPipelin
     // draw's descriptor set.
     descriptor_command_binding.ssbo_block.CreateBuffer(loc, &buffer_info, &alloc_info);
 
-    auto ssbo_memory = (glsl::DescriptorStateSSBO *)descriptor_command_binding.ssbo_block.MapMemory(loc);
```
Comment (spencer-lunarg) [inline layers/gpu/descriptor_validation/gpuav_descriptor_validation.cpp:95]: more renaming can be done later, did some here, naming will never be "perfect" and its is a step forward and not the main focus of this PR

File: tests/spirv/instrumentation.cpp [hunk 3]
```diff
@@ -21,6 +21,15 @@ static constexpr uint32_t kDefaultShaderId = 23;
 static constexpr uint32_t kInstDefaultDescriptorSet = 3;
 static constexpr uint32_t kInstDefaultDebugPrintfBinding = 0;
 
+// While desireable for the instrumentation to be agnostic of the incoming pipeline, we do need to know how the descriptors are laid
```
Comment (spencer-lunarg) [inline tests/spirv/instrumentation.cpp:24]: This is the one "downside" of doing this, from here on forward, the standalone tool will need to know which descriptor sets are used... the old `spirv-opt` has the same issue, it is inevitable and why you can "just compiler a run" a fragment shader from just SPIR-V

... This is mainly for debugging, so for the short term, ok with this approach (in the future, we could use spirv-reflect to automatically detect this)

File: layers/gpu/shaders/instrumentation/bindless_descriptor.comp [hunk 4]
```diff
@@ -68,26 +54,23 @@ layout(set = kInstDefaultDescriptorSet, binding = kBindingInstBindlessDescriptor
 } bindless_state_buffer;
 
 bool inst_bindless_descriptor(const uint inst_num, const uvec4 stage_info, const uint desc_set,
-                              const uint binding, const uint desc_index, const uint byte_offset) {
+                              const uint binding, const uint desc_index, const uint byte_offset, const uint binidng_layout_size, const uint binidng_layout_offset) {
```
Comment (jeremyg-lunarg) [inline layers/gpu/shaders/instrumentation/bindless_descriptor.comp:57]: bad argument name: `binidng_layout_size`

File: layers/gpu/shaders/instrumentation/bindless_descriptor.comp [hunk 5]
```diff
@@ -106,7 +89,7 @@ bool inst_bindless_descriptor(const uint inst_num, const uvec4 stage_info, const
         }
         if (desc_type == kImageSamplerDesc) {
             // check sampler
-            desc_id = ds_type.data[state_index].y;
+            desc_id = ds_type.data[binding_layout_index].y;
```
Comment (arno-lunarg) [inline layers/gpu/shaders/instrumentation/bindless_descriptor.comp:92]: image

**Review by arno-lunarg — APPROVED** [review]
Please do the renaming in another PR so I can more easily have a look!

File: layers/gpu/descriptor_validation/gpuav_descriptor_set.cpp [hunk 6]
```diff
@@ -37,70 +37,31 @@ DescriptorSet::DescriptorSet(const VkDescriptorSet handle, vvl::DescriptorPool *
...
-            state_start += binding->count;
+            binding_layouts_[binding->binding] = {start, binding->count};
+            start += binding->count;
         }
     }
-
-    layout_block_.FlushAllocation(loc);
-    layout_block_.UnmapMemory();
-    return layout_block_.Address();
+    total_descriptor_count_ = start;
```
[the remark this answers is not in the document]
Comment (spencer-lunarg) [reply layers/gpu/descriptor_validation/gpuav_descriptor_set.cpp:64]: We have a `GetTotalDescriptorCount()` function, I wanted to use it, but couldn't confirm it was what I wanted... so I wanted to keep the name the same

File: layers/gpu/instrumentation/gpuav_shader_instrumentor.h [hunk 7]
```diff
@@ -179,8 +180,23 @@ class GpuShaderInstrumentor : public ValidationStateTracker {
...
+
+    // When instrumenting, we need information about the array of VkDescriptorSetLayouts. The core issue is for pipelines, we might
+    // have to merge 2 pipeline layouts together (because of GPL) and therefor both ShaderObject and PipelineLayout state objects
+    // don't have a single way to describe their VkDescriptorSetLayouts. If there are multiple shaders, we also want to only build
+    // this information once. This struct is designed to be filled in from both Pipeline and ShaderObject and then passed down to
+    // the SPIR-V Instrumentation, and afterwards we don't need to save it.
+    struct InstrumentationDescriptorSetLayouts {
+        bool has_bindless_descriptors = false;
+        // < set , [ bindings ] >
+        std::vector<std::vector<spirv::BindingLayout>> binding_layout_lut;
```
[the remark this answers is not in the document]
Comment (spencer-lunarg) [reply layers/gpu/instrumentation/gpuav_shader_instrumentor.h:192]: I feel we don't do this anywhere else, but will do here

File: layers/gpu/resources/gpuav_shader_resources.h [hunk 8]
```diff
@@ -52,24 +52,32 @@ namespace glsl {
...
 struct BindingLayout {
+    uint32_t start;
     uint32_t count;
-    uint32_t state_start;
 };
 
+// Each indexing into a descriptor we have a 32-bit slot to mark what happend on the GPU
+// Since most devices can only support 32 descriptor sets, we can start to be clever and compress info into these 32-bits by
+// exploiting the fact certain data we are saving doesn't need a full 32-bit to save. GLSL doesn't have bitfields and don't want to
+// make this structs until we can't fit everything in 32-bits anymore.
```
[the remark this answers is not in the document]
Comment (jeremyg-lunarg) [reply layers/gpu/resources/gpuav_shader_resources.h:78]: happent -> happened

File: layers/gpu/instrumentation/gpuav_shader_instrumentor.cpp [hunk 9]
```diff
@@ -790,35 +791,55 @@ bool GpuShaderInstrumentor::NeedPipelineCreationShaderInstrumentation(vvl::Pipel
...
-            for (uint32_t i = 0; i < set_layout->GetBindingCount(); i++) {
-                const VkDescriptorBindingFlags flags = set_layout->GetDescriptorBindingFlagsFromIndex(i);
-                if (vvl::IsBindless(flags)) {
-                    return true;
-                }
-            }
+    if (!pipeline_layout) return;
+
+    out_instrumentation_dsl.binding_layout_lut.resize(pipeline_layout->set_layouts.size());
+    for (uint32_t set_layout_index = 0; set_layout_index < pipeline_layout->set_layouts.size(); set_layout_index++) {
```
[the remark this answers is not in the document]
Comment (spencer-lunarg) [reply layers/gpu/instrumentation/gpuav_shader_instrumentor.cpp:800]: we use `set_layout_index` below in the functions, to not use it here is just inconsistent

File: layers/gpu/instrumentation/gpuav_shader_instrumentor.cpp [hunk 10]
```diff
@@ -790,35 +791,55 @@ bool GpuShaderInstrumentor::NeedPipelineCreationShaderInstrumentation(vvl::Pipel
...
+                                                         InstrumentationDescriptorSetLayouts &out_instrumentation_dsl) {
+    if (set_layout_state.GetBindingCount() == 0) return;
+    const uint32_t binding_count = set_layout_state.GetMaxBinding() + 1;
+
+    auto &binding_layouts = out_instrumentation_dsl.binding_layout_lut[set_layout_index];
+    binding_layouts.resize(binding_count);
+
+    uint32_t start = 0;
+    auto dsl_bindings = set_layout_state.GetBindings();
+    for (uint32_t binding_index = 0; binding_index < dsl_bindings.size(); binding_index++) {
```
[the remark this answers is not in the document]
Comment (spencer-lunarg) [reply layers/gpu/instrumentation/gpuav_shader_instrumentor.cpp:828]: its a vector, am I suppose to use it with `data/size`? we don't do that anywhere else I know of

File: layers/gpu/instrumentation/gpuav_shader_instrumentor.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (spencer-lunarg) [reply layers/gpu/instrumentation/gpuav_shader_instrumentor.h:192]: added, makes things harder to read IMO, but updated

**Review by arno-lunarg — APPROVED** [review]
Thank you for the renaming!
Looking at the code close to what you changed, next items on the renaming list should at least be:
- `descriptor_command_binding.ssbo_block`
ssbo is too generic
- `DeviceMemoryBlock`  from `gpuav_resources.h`
Worst name I could come up with for a class that just wraps a VkBuffer. Not only that, but Vma already defines a `VmaDeviceMemoryBlock`. 
But I have no good idea for a new name. Ideally it would be something like `WrappedVkBuffer`, `BufferObject`, but I don't like those

Also worth re-reading Google style guide on Naming Rules: https://google.github.io/styleguide/cppguide.html#General_Naming_Rules

**Pull request merged**